.github/workflows/CI-test-publish.yml

```diff
@@ -0,0 +1,82 @@
+name: Test and Publish Docker Image
+
+on:
+  workflow_call:
+    inputs:
+      artifact_name:
+        description: "Name of artifact of Docker image"
+        required: true
+        type: string
+      git_tag:
+        type: string
+    secrets:
+      DOCKER_HUB_USERNAME:
+        required: true
+      DOCKER_HUB_TOKEN:
+        required: true
+
+jobs:
+  execute-e2e-tests:
+    runs-on: ubuntu-latest
+    name: Execute E2E Tests
+    outputs:
+      docker_image_tag: ${{ steps.set-docker-tag.outputs.docker_image_tag }}
+      test_status: ${{ steps.runtest.outputs.test_status }} 
+    steps:
+      - name: E2E
+        run: echo "execute-e2e-tests with ${{ inputs.artifact_name }}"  
+
+      - name: Download zkVerify Docker image artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: ${{ inputs.artifact_name }}
+          path: ./
+
+      - name: Load Docker image
+        run: |
+          image_name="$(docker load -i "${GITHUB_WORKSPACE}/${image_artifact}.tar" | awk '/Loaded image:/ { print $3 }')"
+          log_info "=== Loaded image ${image_name} ==="
+        shell: bash
+
+      - name: Run test
+        id: run-test
+        uses: 'HorizenLabs/zkVerify-qa/.github/actions/e2e-test@main'
+        with:
+          artifact_name: ${{ inputs.artifact_name }}
+          zkverify_docker_tag: latest
+          GH_TOKEN: ${{ secrets.GH_TOKEN }}
+
+  publish-docker-image:
+    name: Publish Docker Image
+    runs-on: ubuntu-latest
+    # needs: execute-e2e-tests # Wait for e2e tests to finish
+    # if: ${{ needs.execute-e2e-tests.outputs.test_status == 'success' }}
+    steps:
+      - name: Git checkout
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.git_tag || github.ref }}
+          fetch-depth: 0
+
+      - name: Download zkVerify Docker image artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: ${{ inputs.artifact_name }}
+          path: ./
+
+      - name: Publish Docker image
+        env:
+          RELEASE_BRANCH: release
+          MAINTAINERS_KEYS: ${{ vars.MAINTAINERS_KEYS }}
+          DOCKER_HUB_USERNAME: ${{ secrets.DOCKER_HUB_USERNAME }}
+          DOCKER_HUB_TOKEN: ${{ secrets.DOCKER_HUB_TOKEN }}
+        run: |
+          source "${GITHUB_WORKSPACE}/ci/setup_env.sh"
+          "${GITHUB_WORKSPACE}/ci/docker.sh" --image-artifact ${{ inputs.artifact_name }}
+        shell: bash
+
+      - name: Upload runtime artifact
```

Comment:
Is this workflow called after the tag orchestrator one ? That's why you moved it here ?

Reply:
It's called from the CI-tag-orchestrator now.

There is a matrix that runs build, test, and publish each for solo and relay chain
https://github.com/zkVerify/zkVerify/pull/171/files#diff-f93eb3b00b22939661e1b3f2100a5684cdea4aeb8580b8814d0b2780025d6ef9R27-R35